File: fortran90/day2_part1_2020.f90
program password_validator
    implicit none
    integer :: i, min, max, count, valid_count
    character(len=1) :: char
    character(len=256) :: policy, password, line, policy_parts(3)
    open(unit=10, file='input.txt', status='old', action='read')

    valid_count = 0
    do
        read(10, '(A)', end=10) line
        i = index(line, ':') - 1
        policy = line(:i)
        password = line(i+2:)

        i = index(policy, '-')
        policy_parts(1) = policy(:i-1)
        policy_parts(2) = policy(i+1:index(policy, ' ')-1)
        policy_parts(3) = policy(index(policy, ' ')+1:)

        read(policy_parts(1), *) min
        read(policy_parts(2), *) max
        char = policy_parts(3)

        count = 0
        do i = 1, len_trim(password)
            if (password(i:i) == char) count = count + 1
        end do
        if (count >= min .and. count <= max) valid_count = valid_count + 1
    end do
10  close(10)
    print *, valid_count
end program password_validator

File: objectivec/day2_part1_2020.m
#import <Foundation/Foundation.h>

BOOL validatePassword(NSString *policy, NSString *password) {
    NSArray *parts = [policy componentsSeparatedByString:@" "];
    NSArray *range = [[parts firstObject] componentsSeparatedByString:@"-"];
    NSInteger min = [range[0] integerValue];
    NSInteger max = [range[1] integerValue];
    unichar character = [parts[1] characterAtIndex:0];
    
    NSInteger count = 0;
    for (NSUInteger i = 0; i < password.length; i++) {
        if ([password characterAtIndex:i] == character) count++;
    }
    return count >= min && count <= max;
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSInteger validCount = 0;
        NSString *filePath = @"input.txt";
        NSError *error = nil;
        NSString *fileContents = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:&error];
        
        if (error) {
            return 1;
        }
        
        NSArray *lines = [fileContents componentsSeparatedByString:@"\n"];
        for (NSString *line in lines) {
            if ([line length] > 0) {
                NSRange colonRange = [line rangeOfString:@":"];
                if (colonRange.location != NSNotFound) {
                    NSString *policy = [line substringToIndex:colonRange.location];
                    NSString *password = [line substringFromIndex:colonRange.location + 2];
                    if (validatePassword(policy, password)) {
                        validCount++;
                    }
                }
            }
        }
        
        NSLog(@"%ld", (long)validCount);
    }
    return 0;
}